x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/SecurityContext.java

```diff
@@ -186,6 +204,24 @@ public void executeAfterRewritingAuthentication(Consumer<StoredContext> consumer
         }
     }
 
+    /**
+     * Executes consumer in a new thread context after removing {@link AuthorizationEngine.ParentActionAuthorization}.
+     * The original context is provided to the consumer. When this method returns, the original context is restored.
+     */
+    public void executeAfterRemovingParentAuthorization(Consumer<StoredContext> consumer) {
+        final Map<String, String> existingRequestHeaders = threadContext.getRequestHeadersOnly();
+        final StoredContext original = threadContext.newStoredContextPreservingResponseHeaders();
+        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+            existingRequestHeaders.forEach((k, v) -> {
+                if (AuthorizationEngine.ParentActionAuthorization.THREAD_CONTEXT_KEY.equals(k) == false
+                    && threadContext.getHeader(k) == null) {
+                    threadContext.putHeader(k, v);
+                }
+            });
```

Comment:
This does more than clearing the parent authz header. It also resets transient headers.

If I'm right, I'm not sure how the code in `SecurityServerTransportInterceptor#sendRequestInner` can still work correctly, because it inspects transient values that this method clears.

Reply:
Changed remove method in the commit https://github.com/elastic/elasticsearch/pull/91886/commits/db6cc06232db00196cef1ae6f1162340dd0f870f and adjusted `ThreadContext#getTransientHeaders` in https://github.com/elastic/elasticsearch/pull/91886/commits/091612556038a80c443b3cceb83d6c8ddacd9683.

Reply:
It still clears the `ThreadContextStruct#isSystemContext` and the `warningHeadersSize`.
There is a method `ThreadContext#newStoredContext` that clears specified transient headers.
Maybe you could adapt that to take another parameter for the persistent headers to clear, and then reuse that method here in `SecurityContext`?

Reply:
Sorry for this slip again and thanks for the suggestion. I definitely need to make a full copy of the `ThreadContextStruct` and `ThreadContext#newStoredContext` looks fitting since is only ever used in `AuthorizationService`.

Reply:
I've adapted `ThreadContext#newStoredContext` and added additional collection for request headers.
I also introduced new `ThreadContext#newStoredContextPreservingResponseHeaders` method which does the same (clears given transient and request headers) but also preserves response headers.

https://github.com/elastic/elasticsearch/pull/91886/commits/455e0dfcc125dbf6ced20142dba6fa1edede52b0